File: backend/app/schemas/logistics.py
"""
Pydantic v2 schemas for logistics API responses & smart logistics telemetry.
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


# ── Position / Truck ──────────────────────────────────────────────

class TruckPositionResponse(BaseModel):
    """Live truck position, telemetry, and ETA data."""

    id: str
    truck_code: str
    trailer_id: str
    driver_name: Optional[str] = None
    status: str
    priority: str = "NORMAL"
    load_type: Optional[str] = None

    current_lat: float
    current_lng: float
    display_lat: Optional[float] = None
    display_lng: Optional[float] = None
    progress_percent: int = 0

    origin_name: Optional[str] = None
    dest_name: Optional[str] = None
    origin_lat: Optional[float] = None
    origin_lng: Optional[float] = None
    dest_lat: Optional[float] = None
    dest_lng: Optional[float] = None

    original_eta: Optional[datetime] = None
    current_eta: Optional[datetime] = None
    delay_minutes: int = 0

    shipment_id: Optional[str] = None
    updated_at: Optional[datetime] = None

    # Telemetry fields from Smart Logistics Dataset
    source_asset_id: Optional[str] = None
    inventory_level: Optional[float] = None
    temperature: Optional[float] = None
    humidity: Optional[float] = None
    traffic_status: Optional[str] = None
    waiting_time: Optional[float] = None
    logistics_delay_reason: Optional[str] = None
    asset_utilization: Optional[float] = None
    demand_forecast: Optional[float] = None
    is_delayed: Optional[bool] = False
    latest_telemetry_timestamp: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ── Telemetry History ─────────────────────────────────────────────

class TruckTelemetryResponse(BaseModel):
    """Time-series telemetry event record."""

    id: str
    truck_id: Optional[str] = None
    source_timestamp: datetime
    source_asset_id: str
    source_latitude: float
    source_longitude: float
    display_latitude: Optional[float] = None
    display_longitude: Optional[float] = None
    inventory_level: Optional[float] = None
    shipment_status: Optional[str] = None
    temperature: Optional[float] = None
    humidity: Optional[float] = None
    traffic_status: Optional[str] = None
    waiting_time: Optional[float] = None
    logistics_delay_reason: Optional[str] = None
    asset_utilization: Optional[float] = None
    demand_forecast: Optional[float] = None
    logistics_delay: bool = False
    user_transaction_amount: Optional[float] = None
    user_purchase_frequency: Optional[float] = None
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ── Alert ─────────────────────────────────────────────────────────

class LogisticsAlertResponse(BaseModel):
    """An unresolved operational alert."""

    id: str
    truck_id: Optional[str] = None
    alert_type: str
    severity: str
    message: str
    is_resolved: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Shipment ──────────────────────────────────────────────────────

class ShipmentInfoResponse(BaseModel):
    """Shipment header information with resolved corridor coordinates."""

    id: str
    shipment_code: str
    tracking_number: str
    purchase_order_reference: str
    origin_location: str
    destination_location: str
    origin_state: Optional[str] = None
    dest_state: Optional[str] = None
    origin_lat: Optional[float] = None
    origin_lng: Optional[float] = None
    dest_lat: Optional[float] = None
    dest_lng: Optional[float] = None
    status: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ── Route Coordinates ─────────────────────────────────────────────

class RoutePoint(BaseModel):
    """A geo-coordinate point on the route corridor."""
    lat: float
    lng: float
    name: Optional[str] = None


# ── Composite tracking response ───────────────────────────────────

class TrackingSearchResponse(BaseModel):
    """Full tracking result: shipment + truck + alerts + real road route."""

    shipment: Optional[ShipmentInfoResponse] = None
    truck: Optional[TruckPositionResponse] = None
    alerts: list[LogisticsAlertResponse] = Field(default_factory=list)
    route: list[RoutePoint] = Field(default_factory=list)
    route_waypoints: list[list[float]] = Field(default_factory=list)
    corridor_name: Optional[str] = None
    distance_km: Optional[float] = None
    eta: Optional[str] = None
    original_eta: Optional[str] = None
    is_delayed: bool = False
    status: Optional[str] = None


# ── Analytics Summary ─────────────────────────────────────────────

class LogisticsAnalyticsSummaryResponse(BaseModel):
    """Aggregated operational intelligence from smart logistics dataset."""

    total_tracked_assets: int
    in_transit_count: int
    delivered_count: int
    delayed_count: int
    logistics_delay_rate: float
    average_waiting_time: float
    average_asset_utilization: float
    average_temperature: float
    average_humidity: float
    traffic_status_distribution: Dict[str, int]
    delay_reason_distribution: Dict[str, int]
    total_inventory: float
    average_demand_forecast: float
    total_transaction_amount: float
    average_purchase_frequency: float

File: frontend/src/types/logistics.ts
export interface ShipmentInfo {
  id: string;
  shipment_code: string;
  tracking_number: string;
  purchase_order_reference: string;
  origin_location: string;
  destination_location: string;
  origin_state?: string | null;
  dest_state?: string | null;
  origin_lat?: number | null;
  origin_lng?: number | null;
  dest_lat?: number | null;
  dest_lng?: number | null;
  status: string;
  created_at?: string;
  updated_at?: string;
  shipment?: ShipmentInfo | null;
}

export interface RoutePoint {
  lat: number;
  lng: number;
  name?: string | null;
}

export interface TruckPosition {
  id: string;
  truck_code: string;
  trailer_id: string;
  driver_name: string | null;
  status: string;
  priority: string;
  load_type: string | null;
  current_lat: number;
  current_lng: number;
  display_lat?: number | null;
  display_lng?: number | null;
  origin_name?: string | null;
  dest_name?: string | null;
  origin_state?: string | null;
  dest_state?: string | null;
  origin_lat?: number | null;
  origin_lng?: number | null;
  dest_lat?: number | null;
  dest_lng?: number | null;
  progress_percent: number;
  original_eta: string | null;
  current_eta: string | null;
  delay_minutes: number;
  shipment_id?: string | null;
  updated_at?: string;
  shipment?: ShipmentInfo | null;

  // Smart Logistics Telemetry
  source_asset_id?: string | null;
  inventory_level?: number | null;
  temperature?: number | null;
  humidity?: number | null;
  traffic_status?: string | null;
  waiting_time?: number | null;
  logistics_delay_reason?: string | null;
  asset_utilization?: number | null;
  demand_forecast?: number | null;
  is_delayed?: boolean;
  latest_telemetry_timestamp?: string | null;
}

export interface TruckTelemetry {
  id: string;
  truck_id?: string | null;
  source_timestamp: string;
  source_asset_id: string;
  source_latitude: number;
  source_longitude: number;
  display_latitude?: number | null;
  display_longitude?: number | null;
  inventory_level?: number | null;
  shipment_status?: string | null;
  temperature?: number | null;
  humidity?: number | null;
  traffic_status?: string | null;
  waiting_time?: number | null;
  logistics_delay_reason?: string | null;
  asset_utilization?: number | null;
  demand_forecast?: number | null;
  logistics_delay: boolean;
  user_transaction_amount?: number | null;
  user_purchase_frequency?: number | null;
  created_at?: string;
}

export interface LogisticsAlert {
  id: string;
  truck_id: string;
  alert_type: string;
  severity: "INFO" | "WARNING" | "CRITICAL";
  message: string;
  is_resolved: boolean;
  created_at: string;
}

export interface TrackingSearchResponse {
  truck: TruckPosition;
  shipment?: ShipmentInfo | null;
  route?: RoutePoint[];
  route_waypoints?: [number, number][];
  corridor_name?: string | null;
  distance_km?: number | null;
  alerts: LogisticsAlert[];
  eta?: string | null;
  original_eta?: string | null;
  is_delayed?: boolean;
  status?: string;
}

export interface LogisticsAnalyticsSummary {
  total_tracked_assets: number;
  in_transit_count: number;
  delivered_count: number;
  delayed_count: number;
  logistics_delay_rate: number;
  average_waiting_time: number;
  average_asset_utilization: number;
  average_temperature: number;
  average_humidity: number;
  traffic_status_distribution: Record<string, number>;
  delay_reason_distribution: Record<string, number>;
  total_inventory: number;
  average_demand_forecast: number;
  total_transaction_amount: number;
  average_purchase_frequency: number;
}

export interface YardSlotResponse {
  id: string;
  slot_code: string;
  status: string;
  truck_id: string | null;
  appointment_time: string | null;
  created_at: string;
  updated_at: string;
  truck?: TruckPosition | null;
}

export interface DockResponse {
  id: string;
  dock_code: string;
  status: string;
  suitable_load_types: string;
  current_truck_id: string | null;
  created_at: string;
  updated_at: string;
  current_truck?: TruckPosition | null;
}

export interface DockAssignmentResponse {
  id: string;
  truck_id: string;
  dock_id: string;
  recommended_score: number;
  status: string;
  assigned_at: string | null;
  departed_at: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
  truck?: TruckPosition | null;
  dock?: DockResponse | null;
}

export interface DockRecommendationResponse {
  recommended_dock: DockResponse | null;
  score: number;
  reason: string;
  alternatives: DockResponse[];
}

export type DockAlertResponse = LogisticsAlert;

